gpu-socket feature causing feature unification issues

since the set_gpu_socket feature is behind a feature flag, this cause feature unification issues, because vhost-device-gpu is enabling that feature and other crates are not enabling it. According to rust feature reference: 

> That is, enabling a feature should not disable functionality, and it should usually be safe to enable any combination of features. A feature should not introduce a [SemVer-incompatible change](https://doc.rust-lang.org/cargo/reference/features.html#semver-compatibility).

Just a bit more context. The issue was discovered while adding `vhost-user-gpu` crate in a workspace with other crates using the `vhost-user-backend` crate, see https://github.com/rust-vmm/vhost-device/pull/668

Since the new crate enable `gpu-socket` feature, this is enabled by cargo also for other crates in the workspace to minimize the number of copy of the crate, see https://doc.rust-lang.org/cargo/reference/features.html#feature-unification

At this point, we had the following issue on another crate in the same workspace (`vhost-device-console`) not enabling that feature:
```
   Compiling vhost-device-console v0.1.0 (/workdir/staging/vhost-device-console)
error[E0046]: not all trait items implemented, missing: `set_gpu_socket`
   --> vhost-device-console/src/vhu_console.rs:641:1
    |
641 | impl VhostUserBackendMut for VhostUserConsoleBackend {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `set_gpu_socket` in implementation
```

The first fix we had in mind was in some way to avoid the feature unification, maybe forcing cargo to use 2 copies, but at that point we discovered the recommendation in https://doc.rust-lang.org/cargo/reference/features.html#feature-unification that also @dorindabassey reported:

> That is, enabling a feature should not disable functionality, and it should usually be safe to enable any combination of features. A feature should not introduce a [SemVer-incompatible change](https://doc.rust-lang.org/cargo/reference/features.html#semver-compatibility).

So maybe we should avoid that here, so I was thinking of not really removing the feature, but something like just providing default implementation for VhostUserBackend that should not be used if the feature is not enabled (maybe we can panic there):
```diff
diff --git a/vhost-user-backend/src/backend.rs b/vhost-user-backend/src/backend.rs
index ad9e950..2ba7576 100644
--- a/vhost-user-backend/src/backend.rs
+++ b/vhost-user-backend/src/backend.rs
@@ -87,13 +87,12 @@ pub trait VhostUserBackend: Send + Sync {
     /// function.
     fn set_backend_req_fd(&self, _backend: Backend) {}
 
-    #[cfg(feature = "gpu-socket")]
     /// Set handler for communicating with the frontend by the gpu specific backend communication
     /// channel.
     ///
     /// This method only exits when the crate feature gpu-socket is enabled, because this is only
     /// useful for a gpu device.
-    fn set_gpu_socket(&self, _gpu_backend: GpuBackend);
+    fn set_gpu_socket(&self, _gpu_backend: GpuBackend) {}
 
     /// Get the map to map queue index to worker thread index.
     ///
(END)
```

@germag @epilys any suggestion?

--- a/.buildkite/rust-vmm-ci-tests.json
+++ b/.buildkite/rust-vmm-ci-tests.json
@@ -22,7 +22,7 @@
     },
     {
       "test_name": "unittests-gnu-all-with-xen",
-      "command": "cargo test --workspace --no-default-features --features test-utils,vhost-vsock,vhost-kern,vhost-vdpa,vhost-net,vhost-user,vhost-user-frontend,gpu-socket,vhost-user-backend,xen",
+      "command": "cargo test --workspace --no-default-features --features test-utils,vhost-vsock,vhost-kern,vhost-vdpa,vhost-net,vhost-user,vhost-user-frontend,vhost-user-backend,xen",
       "platform": [
         "x86_64",
         "aarch64"
@@ -30,7 +30,7 @@
     },
     {
       "test_name": "unittests-gnu-all-without-xen",
-      "command": "cargo test --workspace --no-default-features --features test-utils,vhost-vsock,vhost-kern,vhost-vdpa,vhost-net,vhost-user,vhost-user-frontend,gpu-socket,vhost-user-backend,postcopy",
+      "command": "cargo test --workspace --no-default-features --features test-utils,vhost-vsock,vhost-kern,vhost-vdpa,vhost-net,vhost-user,vhost-user-frontend,vhost-user-backend,postcopy",
       "platform": [
         "x86_64",
         "aarch64"
@@ -38,7 +38,7 @@
     },
     {
       "test_name": "unittests-musl-all-with-xen",
-      "command": "cargo test --workspace --target {target_platform}-unknown-linux-musl --no-default-features --features test-utils,vhost-vsock,vhost-kern,vhost-vdpa,vhost-net,vhost-user,vhost-user-frontend,gpu-socket,vhost-user-backend,xen",
+      "command": "cargo test --workspace --target {target_platform}-unknown-linux-musl --no-default-features --features test-utils,vhost-vsock,vhost-kern,vhost-vdpa,vhost-net,vhost-user,vhost-user-frontend,vhost-user-backend,xen",
       "platform": [
         "x86_64",
         "aarch64"
@@ -46,7 +46,7 @@
     },
     {
       "test_name": "unittests-musl-all-without-xen",
-      "command": "cargo test --workspace --target {target_platform}-unknown-linux-musl --no-default-features --features test-utils,vhost-vsock,vhost-kern,vhost-vdpa,vhost-net,vhost-user,vhost-user-frontend,gpu-socket,vhost-user-backend,postcopy",
+      "command": "cargo test --workspace --target {target_platform}-unknown-linux-musl --no-default-features --features test-utils,vhost-vsock,vhost-kern,vhost-vdpa,vhost-net,vhost-user,vhost-user-frontend,vhost-user-backend,postcopy",
       "platform": [
         "x86_64",
         "aarch64"
@@ -54,7 +54,7 @@
     },
     {
       "test_name": "clippy-all-with-xen",
-      "command": "cargo clippy --workspace --bins --examples --benches --all-targets --no-default-features --features test-utils,vhost-vsock,vhost-kern,vhost-vdpa,vhost-net,vhost-user,vhost-user-frontend,gpu-socket,vhost-user-backend,xen -- -D warnings -D clippy::undocumented_unsafe_blocks",
+      "command": "cargo clippy --workspace --bins --examples --benches --all-targets --no-default-features --features test-utils,vhost-vsock,vhost-kern,vhost-vdpa,vhost-net,vhost-user,vhost-user-frontend,vhost-user-backend,xen -- -D warnings -D clippy::undocumented_unsafe_blocks",
       "platform": [
         "x86_64",
         "aarch64"
@@ -62,7 +62,7 @@
     },
     {
       "test_name": "clippy-all-without-xen",
-      "command": "cargo clippy --workspace --bins --examples --benches --all-targets --no-default-features --features test-utils,vhost-vsock,vhost-kern,vhost-vdpa,vhost-net,vhost-user,vhost-user-frontend,gpu-socket,vhost-user-backend,postcopy -- -D warnings -D clippy::undocumented_unsafe_blocks",
+      "command": "cargo clippy --workspace --bins --examples --benches --all-targets --no-default-features --features test-utils,vhost-vsock,vhost-kern,vhost-vdpa,vhost-net,vhost-user,vhost-user-frontend,vhost-user-backend,postcopy -- -D warnings -D clippy::undocumented_unsafe_blocks",
       "platform": [
         "x86_64",
         "aarch64"
@@ -70,7 +70,7 @@
     },
     {
       "test_name": "check-warnings-all-with-xen",
-      "command": "RUSTFLAGS=\"-D warnings\" cargo check --all-targets --workspace --no-default-features --features test-utils,vhost-vsock,vhost-kern,vhost-vdpa,vhost-net,vhost-user,vhost-user-frontend,gpu-socket,vhost-user-backend,xen",
+      "command": "RUSTFLAGS=\"-D warnings\" cargo check --all-targets --workspace --no-default-features --features test-utils,vhost-vsock,vhost-kern,vhost-vdpa,vhost-net,vhost-user,vhost-user-frontend,vhost-user-backend,xen",
       "platform": [
         "x86_64",
         "aarch64"
@@ -78,7 +78,7 @@
     },
     {
       "test_name": "check-warnings-all-without-xen",
-      "command": "RUSTFLAGS=\"-D warnings\" cargo check --all-targets --workspace --no-default-features --features test-utils,vhost-vsock,vhost-kern,vhost-vdpa,vhost-net,vhost-user,vhost-user-frontend,gpu-socket,vhost-user-backend,postcopy",
+      "command": "RUSTFLAGS=\"-D warnings\" cargo check --all-targets --workspace --no-default-features --features test-utils,vhost-vsock,vhost-kern,vhost-vdpa,vhost-net,vhost-user,vhost-user-frontend,vhost-user-backend,postcopy",
       "platform": [
         "x86_64",
         "aarch64"

--- a/vhost-user-backend/CHANGELOG.md
+++ b/vhost-user-backend/CHANGELOG.md
@@ -8,6 +8,7 @@
 ### Deprecated
 
 ### Fixed
+- [[#267](https://github.com/rust-vmm/vhost/pull/267)] Fix feature unification issues with gpu-socket feature.
 
 ## v0.16.0
 

--- a/vhost-user-backend/Cargo.toml
+++ b/vhost-user-backend/Cargo.toml
@@ -11,7 +11,6 @@ license = "Apache-2.0"
 [features]
 xen = ["vm-memory/xen", "vhost/xen"]
 postcopy = ["vhost/postcopy", "userfaultfd"]
-gpu-socket = ["vhost/gpu-socket"]
 
 [dependencies]
 libc = "0.2.39"

--- a/vhost-user-backend/src/backend.rs
+++ b/vhost-user-backend/src/backend.rs
@@ -31,7 +31,6 @@ use vm_memory::bitmap::Bitmap;
 use vmm_sys_util::epoll::EventSet;
 use vmm_sys_util::eventfd::EventFd;
 
-#[cfg(feature = "gpu-socket")]
 use vhost::vhost_user::GpuBackend;
 
 use super::vring::VringT;
@@ -87,7 +86,6 @@ pub trait VhostUserBackend: Send + Sync {
     /// function.
     fn set_backend_req_fd(&self, _backend: Backend) {}
 
-    #[cfg(feature = "gpu-socket")]
     /// Set handler for communicating with the frontend by the gpu specific backend communication
     /// channel.
     ///
@@ -205,7 +203,6 @@ pub trait VhostUserBackendMut: Send + Sync {
     /// function.
     fn set_backend_req_fd(&mut self, _backend: Backend) {}
 
-    #[cfg(feature = "gpu-socket")]
     /// Set handler for communicating with the frontend by the gpu specific backend communication
     /// channel.
     ///
@@ -318,7 +315,6 @@ impl<T: VhostUserBackend> VhostUserBackend for Arc<T> {
         self.deref().set_backend_req_fd(backend)
     }
 
-    #[cfg(feature = "gpu-socket")]
     fn set_gpu_socket(&self, gpu_backend: GpuBackend) {
         self.deref().set_gpu_socket(gpu_backend)
     }
@@ -400,7 +396,6 @@ impl<T: VhostUserBackendMut> VhostUserBackend for Mutex<T> {
         self.lock().unwrap().set_backend_req_fd(backend)
     }
 
-    #[cfg(feature = "gpu-socket")]
     fn set_gpu_socket(&self, gpu_backend: GpuBackend) {
         self.lock().unwrap().set_gpu_socket(gpu_backend)
     }
@@ -485,7 +480,6 @@ impl<T: VhostUserBackendMut> VhostUserBackend for RwLock<T> {
         self.write().unwrap().set_backend_req_fd(backend)
     }
 
-    #[cfg(feature = "gpu-socket")]
     fn set_gpu_socket(&self, gpu_backend: GpuBackend) {
         self.write().unwrap().set_gpu_socket(gpu_backend)
     }
@@ -610,7 +604,6 @@ pub mod tests {
 
         fn set_backend_req_fd(&mut self, _backend: Backend) {}
 
-        #[cfg(feature = "gpu-socket")]
         fn set_gpu_socket(&mut self, _gpu_backend: GpuBackend) {}
 
         fn queues_per_thread(&self) -> Vec<u64> {

--- a/vhost-user-backend/src/backend.rs
+++ b/vhost-user-backend/src/backend.rs
@@ -89,9 +89,14 @@ pub trait VhostUserBackend: Send + Sync {
     /// Set handler for communicating with the frontend by the gpu specific backend communication
     /// channel.
     ///
-    /// This method only exits when the crate feature gpu-socket is enabled, because this is only
-    /// useful for a gpu device.
-    fn set_gpu_socket(&self, _gpu_backend: GpuBackend);
+    /// This function returns a `Result`, returning an error if the backend does not implement this
+    /// function.
+    fn set_gpu_socket(&self, _gpu_backend: GpuBackend) -> Result<()> {
+        Err(std::io::Error::new(
+            std::io::ErrorKind::Unsupported,
+            "backend does not support set_gpu_socket() / VHOST_USER_GPU_SET_SOCKET",
+        ))
+    }
 
     /// Get the map to map queue index to worker thread index.
     ///
@@ -206,9 +211,14 @@ pub trait VhostUserBackendMut: Send + Sync {
     /// Set handler for communicating with the frontend by the gpu specific backend communication
     /// channel.
     ///
-    /// This method only exits when the crate feature gpu-socket is enabled, because this is only
-    /// useful for a gpu device.
-    fn set_gpu_socket(&mut self, gpu_backend: GpuBackend);
+    /// This function returns a `Result`, returning an error if the backend does not implement this
+    /// function.
+    fn set_gpu_socket(&mut self, _gpu_backend: GpuBackend) -> Result<()> {
+        Err(std::io::Error::new(
+            std::io::ErrorKind::Unsupported,
+            "backend does not support set_gpu_socket() / VHOST_USER_GPU_SET_SOCKET",
+        ))
+    }
 
     /// Get the map to map queue index to worker thread index.
     ///
@@ -315,7 +325,7 @@ impl<T: VhostUserBackend> VhostUserBackend for Arc<T> {
         self.deref().set_backend_req_fd(backend)
     }
 
-    fn set_gpu_socket(&self, gpu_backend: GpuBackend) {
+    fn set_gpu_socket(&self, gpu_backend: GpuBackend) -> Result<()> {
         self.deref().set_gpu_socket(gpu_backend)
     }
 
@@ -396,7 +406,7 @@ impl<T: VhostUserBackendMut> VhostUserBackend for Mutex<T> {
         self.lock().unwrap().set_backend_req_fd(backend)
     }
 
-    fn set_gpu_socket(&self, gpu_backend: GpuBackend) {
+    fn set_gpu_socket(&self, gpu_backend: GpuBackend) -> Result<()> {
         self.lock().unwrap().set_gpu_socket(gpu_backend)
     }
 
@@ -480,7 +490,7 @@ impl<T: VhostUserBackendMut> VhostUserBackend for RwLock<T> {
         self.write().unwrap().set_backend_req_fd(backend)
     }
 
-    fn set_gpu_socket(&self, gpu_backend: GpuBackend) {
+    fn set_gpu_socket(&self, gpu_backend: GpuBackend) -> Result<()> {
         self.write().unwrap().set_gpu_socket(gpu_backend)
     }
 
@@ -603,8 +613,6 @@ pub mod tests {
         }
 
         fn set_backend_req_fd(&mut self, _backend: Backend) {}
-
-        fn set_gpu_socket(&mut self, _gpu_backend: GpuBackend) {}
 
         fn queues_per_thread(&self) -> Vec<u64> {
             vec![1, 1]

--- a/vhost-user-backend/src/handler.rs
+++ b/vhost-user-backend/src/handler.rs
@@ -555,8 +555,10 @@ where
         self.backend.set_backend_req_fd(backend);
     }
 
-    fn set_gpu_socket(&mut self, gpu_backend: GpuBackend) {
-        self.backend.set_gpu_socket(gpu_backend);
+    fn set_gpu_socket(&mut self, gpu_backend: GpuBackend) -> VhostUserResult<()> {
+        self.backend
+            .set_gpu_socket(gpu_backend)
+            .map_err(VhostUserError::ReqHandlerError)
     }
 
     fn get_inflight_fd(

--- a/vhost-user-backend/src/handler.rs
+++ b/vhost-user-backend/src/handler.rs
@@ -21,7 +21,6 @@ use vhost::vhost_user::message::{
     VhostUserMemoryRegion, VhostUserProtocolFeatures, VhostUserSingleMemoryRegion,
     VhostUserVirtioFeatures, VhostUserVringAddrFlags, VhostUserVringState,
 };
-#[cfg(feature = "gpu-socket")]
 use vhost::vhost_user::GpuBackend;
 use vhost::vhost_user::{
     Backend, Error as VhostUserError, Result as VhostUserResult, VhostUserBackendReqHandlerMut,
@@ -556,7 +555,6 @@ where
         self.backend.set_backend_req_fd(backend);
     }
 
-    #[cfg(feature = "gpu-socket")]
     fn set_gpu_socket(&mut self, gpu_backend: GpuBackend) {
         self.backend.set_gpu_socket(gpu_backend);
     }

--- a/vhost-user-backend/tests/vhost-user-server.rs
+++ b/vhost-user-backend/tests/vhost-user-server.rs
@@ -10,7 +10,6 @@ use std::thread;
 use vhost::vhost_user::message::{
     VhostUserConfigFlags, VhostUserHeaderFlag, VhostUserInflight, VhostUserProtocolFeatures,
 };
-use vhost::vhost_user::GpuBackend;
 use vhost::vhost_user::{Backend, Frontend, Listener, VhostUserFrontend};
 use vhost::{VhostBackend, VhostUserMemoryRegionInfo, VringConfigData};
 use vhost_user_backend::{VhostUserBackendMut, VhostUserDaemon, VringRwLock};
@@ -77,8 +76,6 @@ impl VhostUserBackendMut for MockVhostBackend {
 
         Ok(())
     }
-
-    fn set_gpu_socket(&mut self, _gpu_backend: GpuBackend) {}
 
     fn update_memory(&mut self, atomic_mem: GuestMemoryAtomic<GuestMemoryMmap>) -> Result<()> {
         let mem = atomic_mem.memory();

--- a/vhost-user-backend/tests/vhost-user-server.rs
+++ b/vhost-user-backend/tests/vhost-user-server.rs
@@ -10,7 +10,6 @@ use std::thread;
 use vhost::vhost_user::message::{
     VhostUserConfigFlags, VhostUserHeaderFlag, VhostUserInflight, VhostUserProtocolFeatures,
 };
-#[cfg(feature = "gpu-socket")]
 use vhost::vhost_user::GpuBackend;
 use vhost::vhost_user::{Backend, Frontend, Listener, VhostUserFrontend};
 use vhost::{VhostBackend, VhostUserMemoryRegionInfo, VringConfigData};
@@ -79,7 +78,6 @@ impl VhostUserBackendMut for MockVhostBackend {
         Ok(())
     }
 
-    #[cfg(feature = "gpu-socket")]
     fn set_gpu_socket(&mut self, _gpu_backend: GpuBackend) {}
 
     fn update_memory(&mut self, atomic_mem: GuestMemoryAtomic<GuestMemoryMmap>) -> Result<()> {

--- a/vhost/CHANGELOG.md
+++ b/vhost/CHANGELOG.md
@@ -8,6 +8,7 @@
 ### Deprecated
 
 ### Fixed
+- [[#267](https://github.com/rust-vmm/vhost/pull/267)] Fix feature unification issues with gpu-socket feature.
 
 ## [0.12.0]
 

--- a/vhost/Cargo.toml
+++ b/vhost/Cargo.toml
@@ -23,7 +23,6 @@ vhost-net = ["vhost-kern"]
 vhost-user = []
 vhost-user-frontend = ["vhost-user"]
 vhost-user-backend = ["vhost-user"]
-gpu-socket = ["vhost-user"]
 xen = ["vm-memory/xen"]
 postcopy = []
 

--- a/vhost/src/vhost_user/backend_req_handler.rs
+++ b/vhost/src/vhost_user/backend_req_handler.rs
@@ -66,7 +66,7 @@ pub trait VhostUserBackendReqHandler {
     fn get_config(&self, offset: u32, size: u32, flags: VhostUserConfigFlags) -> Result<Vec<u8>>;
     fn set_config(&self, offset: u32, buf: &[u8], flags: VhostUserConfigFlags) -> Result<()>;
     fn set_backend_req_fd(&self, _backend: Backend) {}
-    fn set_gpu_socket(&self, gpu_backend: GpuBackend);
+    fn set_gpu_socket(&self, _gpu_backend: GpuBackend) -> Result<()>;
     fn get_inflight_fd(&self, inflight: &VhostUserInflight) -> Result<(VhostUserInflight, File)>;
     fn set_inflight_fd(&self, inflight: &VhostUserInflight, file: File) -> Result<()>;
     fn get_max_mem_slots(&self) -> Result<u64>;
@@ -126,7 +126,7 @@ pub trait VhostUserBackendReqHandlerMut {
     ) -> Result<Vec<u8>>;
     fn set_config(&mut self, offset: u32, buf: &[u8], flags: VhostUserConfigFlags) -> Result<()>;
     fn set_backend_req_fd(&mut self, _backend: Backend) {}
-    fn set_gpu_socket(&mut self, _gpu_backend: GpuBackend);
+    fn set_gpu_socket(&mut self, _gpu_backend: GpuBackend) -> Result<()>;
     fn get_inflight_fd(
         &mut self,
         inflight: &VhostUserInflight,
@@ -238,8 +238,8 @@ impl<T: VhostUserBackendReqHandlerMut> VhostUserBackendReqHandler for Mutex<T> {
         self.lock().unwrap().set_backend_req_fd(backend)
     }
 
-    fn set_gpu_socket(&self, gpu_backend: GpuBackend) {
-        self.lock().unwrap().set_gpu_socket(gpu_backend);
+    fn set_gpu_socket(&self, gpu_backend: GpuBackend) -> Result<()> {
+        self.lock().unwrap().set_gpu_socket(gpu_backend)
     }
 
     fn get_inflight_fd(&self, inflight: &VhostUserInflight) -> Result<(VhostUserInflight, File)> {
@@ -814,8 +814,7 @@ impl<S: VhostUserBackendReqHandler> BackendReqHandler<S> {
         // since we have no way to check this. If not, it will fail later.
         let sock = unsafe { UnixStream::from_raw_fd(file.into_raw_fd()) };
         let gpu_backend = GpuBackend::from_stream(sock);
-        self.backend.set_gpu_socket(gpu_backend);
-        Ok(())
+        self.backend.set_gpu_socket(gpu_backend)
     }
 
     fn handle_vring_fd_request(

--- a/vhost/src/vhost_user/backend_req_handler.rs
+++ b/vhost/src/vhost_user/backend_req_handler.rs
@@ -12,7 +12,6 @@ use vm_memory::ByteValued;
 
 use super::backend_req::Backend;
 use super::connection::Endpoint;
-#[cfg(feature = "gpu-socket")]
 use super::gpu_backend_req::GpuBackend;
 use super::message::*;
 use super::{take_single_file, Error, Result};
@@ -67,7 +66,6 @@ pub trait VhostUserBackendReqHandler {
     fn get_config(&self, offset: u32, size: u32, flags: VhostUserConfigFlags) -> Result<Vec<u8>>;
     fn set_config(&self, offset: u32, buf: &[u8], flags: VhostUserConfigFlags) -> Result<()>;
     fn set_backend_req_fd(&self, _backend: Backend) {}
-    #[cfg(feature = "gpu-socket")]
     fn set_gpu_socket(&self, gpu_backend: GpuBackend);
     fn get_inflight_fd(&self, inflight: &VhostUserInflight) -> Result<(VhostUserInflight, File)>;
     fn set_inflight_fd(&self, inflight: &VhostUserInflight, file: File) -> Result<()>;
@@ -128,7 +126,6 @@ pub trait VhostUserBackendReqHandlerMut {
     ) -> Result<Vec<u8>>;
     fn set_config(&mut self, offset: u32, buf: &[u8], flags: VhostUserConfigFlags) -> Result<()>;
     fn set_backend_req_fd(&mut self, _backend: Backend) {}
-    #[cfg(feature = "gpu-socket")]
     fn set_gpu_socket(&mut self, _gpu_backend: GpuBackend);
     fn get_inflight_fd(
         &mut self,
@@ -241,7 +238,6 @@ impl<T: VhostUserBackendReqHandlerMut> VhostUserBackendReqHandler for Mutex<T> {
         self.lock().unwrap().set_backend_req_fd(backend)
     }
 
-    #[cfg(feature = "gpu-socket")]
     fn set_gpu_socket(&self, gpu_backend: GpuBackend) {
         self.lock().unwrap().set_gpu_socket(gpu_backend);
     }
@@ -571,7 +567,6 @@ impl<S: VhostUserBackendReqHandler> BackendReqHandler<S> {
                 let res = self.backend.set_inflight_fd(&msg, file);
                 self.send_ack_message(&hdr, res)?;
             }
-            #[cfg(feature = "gpu-socket")]
             Ok(FrontendReq::GPU_SET_SOCKET) => {
                 let res = self.set_gpu_socket(files);
                 self.send_ack_message(&hdr, res)?;
@@ -812,7 +807,6 @@ impl<S: VhostUserBackendReqHandler> BackendReqHandler<S> {
         Ok(())
     }
 
-    #[cfg(feature = "gpu-socket")]
     fn set_gpu_socket(&mut self, files: Option<Vec<File>>) -> Result<()> {
         let file = take_single_file(files).ok_or(Error::InvalidMessage)?;
         // SAFETY: Safe because we have ownership of the files that were

--- a/vhost/src/vhost_user/dummy_backend.rs
+++ b/vhost/src/vhost_user/dummy_backend.rs
@@ -259,7 +259,9 @@ impl VhostUserBackendReqHandlerMut for DummyBackendReqHandler {
         Ok(())
     }
 
-    fn set_gpu_socket(&mut self, _gpu_backend: GpuBackend) {}
+    fn set_gpu_socket(&mut self, _gpu_backend: GpuBackend) -> Result<()> {
+        Ok(())
+    }
 
     fn get_inflight_fd(
         &mut self,

--- a/vhost/src/vhost_user/dummy_backend.rs
+++ b/vhost/src/vhost_user/dummy_backend.rs
@@ -259,7 +259,6 @@ impl VhostUserBackendReqHandlerMut for DummyBackendReqHandler {
         Ok(())
     }
 
-    #[cfg(feature = "gpu-socket")]
     fn set_gpu_socket(&mut self, _gpu_backend: GpuBackend) {}
 
     fn get_inflight_fd(

--- a/vhost/src/vhost_user/mod.rs
+++ b/vhost/src/vhost_user/mod.rs
@@ -52,11 +52,8 @@ pub use self::backend_req_handler::{
 mod backend_req;
 #[cfg(feature = "vhost-user-backend")]
 pub use self::backend_req::Backend;
-#[cfg(feature = "gpu-socket")]
 mod gpu_backend_req;
-#[cfg(feature = "gpu-socket")]
 pub mod gpu_message;
-#[cfg(feature = "gpu-socket")]
 pub use self::gpu_backend_req::GpuBackend;
 
 /// Errors for vhost-user operations

